src/components/documents/NetpbmEditor/index.tsx

```diff
@@ -0,0 +1,151 @@
+import React from 'react';
+import clsx from 'clsx';
+import { ParserResult } from './types';
+import ImageCanvas from './ImageCanvas';
+import { MetaInit, ModelMeta } from '@tdev-models/documents/NetpbmGrapic';
+import { useFirstMainDocument } from '@tdev-hooks/useFirstMainDocument';
+import { Source } from '@tdev-models/iDocument';
+import SyncStatus from '@tdev-components/SyncStatus';
+import Icon from '@mdi/react';
+import { mdiAlertCircle, mdiAlertCircleOutline, mdiCheckAll, mdiFlashTriangle } from '@mdi/js';
+import { parse } from './parser/parser';
+
+interface Props extends MetaInit {
+    id: string;
+    noEditor?: boolean;
+}
+
+const NetpbmEditor = observer((props: Props) => {
+    const [sanitizedData, setSanitizedData] = React.useState<string>('');
+    const [displayedErrors, setDisplayedErrors] = React.useState<(string | React.ReactElement)[]>([]);
+    const [displayedWarnings, setDisplayedWarnings] = React.useState<(string | React.ReactElement)[]>([]);
+    const [height, setHeight] = React.useState<number>(0);
+    const [width, setWidth] = React.useState<number>(0);
+    const [pixels, setPixels] = React.useState<Uint8ClampedArray>();
+
+    const [meta] = React.useState(new ModelMeta(props));
+    const doc = useFirstMainDocument(props.id, meta);
+
+    React.useEffect(() => {
+        const result = doc.data.imageData
+            .trim()
+            .split('\n')
+            .filter((line) => !line.trim().startsWith('#')) // Remove comments.
+            .join('\n');
+
+        setSanitizedData(result);
+    }, [doc.data.imageData]);
+
+    const resetImageData = () => {
+        setHeight(0);
+        setWidth(0);
+        setPixels(new Uint8ClampedArray());
+    };
+
+    const resetErrorsAndWarnings = () => {
+        setDisplayedErrors([]);
+        setDisplayedWarnings([]);
+    };
+
+    const processParserResult = (result: ParserResult) => {
+        const { imageData, errors, warnings } = result;
+        setDisplayedErrors(errors || []);
+        setDisplayedWarnings(warnings || []);
+        if (imageData) {
+            setHeight(imageData.height);
+            setWidth(imageData.width);
+            setPixels(imageData.pixels);
+        }
+    };
+
+    const render = () => {
+        resetImageData();
+        resetErrorsAndWarnings();
+        processParserResult(parse(sanitizedData));
+    };
+
+    React.useEffect(() => {
+        render();
+    }, [sanitizedData]);
+
+    const StateIcons = () => (
+        <span className={clsx(styles.stateIcons)}>
+            <SyncStatus model={doc} size={0.7} />
+            {doc.root?.isDummy && (
+                <Icon path={mdiFlashTriangle} size={0.7} color="orange" title="Wird nicht gespeichert." />
+            )}
+        </span>
+    );
+
+    const hasErrorsOrWarnings = () => displayedErrors.length > 0 || displayedWarnings.length > 0;
+    const hasWarnings = () => displayedWarnings.length > 0;
+    const hasErrors = () => displayedErrors.length > 0;
```

Comment:
[perf]: use `React.useMemo`: it will recompute only when it's dependencies change

```suggestion
    const { hasErrorsOrWarnings, hasWarnings, hasErrors } = React.useMemo(() => {
        return {
            hasErrorsOrWarnings: displayedErrors.length > 0 || displayedWarnings.length > 0,
            hasWarnings: displayedWarnings.length > 0,
            hasErrors: displayedErrors.length > 0
        };
    }, [displayedErrors, displayedWarnings]);
```

Reply:
TIL – cool feature 😁